### ISSUE TITLE
[types] torch.utils.data.{Dataset, Sampler} are not Sized

### ISSUE THREAD

**Issue by alanhdu** [pytorch/pytorch#47055]
## 🐛 Bug

The base classes for `torch.utils.data.Dataset` and `torch.utils.data.Sampler` do not define a `__len__` method, which means that downstream code relying on PyTorch (with typehints) cannot assume that the class is supposed to have a `__len__` method (adhering to the "Sized" protocol).

There's a fairly lengthy comment about why this is the case:

https://github.com/pytorch/pytorch/blob/dfdc1dbee4809a7433fab5cb6880eb6396e9e59d/torch/utils/data/sampler.py#L27-L52

I admit that I'm not quite sure what this comment means -- in my small explorations, I didn't encounter either problem (e.g. subclass instances work just fine when returning a `NotImplemented`). That said, I'd like to offer two additional potential solutions:

- Make these classes `abc.ABC`s and make `__len__` an abstractmethod:
```python
@abc.abstractmethod
def __len__(self) -> int:
    ...
```
- Make the base classes inherit from `typing.Sized`.

**Comment by SsnL** [comment]
No. We can't make them `abstractmethod` or `Sized`, because `__len__` is **optional**. Not required.

**Comment by SsnL** [comment]
The problem with returning `NotImplemented` is that there is no stack trace and the error message is confusing.

**Comment by SsnL** [comment]
Really `Sampler` is just an `Iterable`, not `Sized`.

**Comment by alanhdu** [comment]
Hm.. if that's the case, would it make sense to have two different base classes for `Dataset` have `IterableDataset` (that only requires `__iter__`) and then have `MapStyleDataset(IterableDataset)` that requires (`__getitem__` and `__len__`)? The current setup (where `Dataset` has `__getitem__` and `IterableDataset` inherits that requirement) seems backwards.

**Comment by SsnL** [comment]
Re dataset: Yeah, that would be ideal, although I don't have ideas to resolve this. The current status is due to the fact that `IterableDataset` was added after `DataSet` and we wanted to still `isinstance(iterable_dataset, Dataset)` and maintain BC.

**Comment by VitalyFedyunin** [comment]
@alanhdu @SsnL 

We are looking into the option to have something similar to:

```python
class MapDataset(Dataset):
  pass
class IterableDataset(Dataset):
  pass
```

and derive them properly in different situations.

However it might not happen before PyTorch 1.10 (~ sooner than 2-4 month)

**Comment by bryant1410** [comment]
What if the Sampler, by default, declares the `__len__` method only if the data source has it? Like:

```python
def __init__(self, data_source: Optional[Any] = None) -> None:
    if hasattr(sampler, "__len__"):
        self.__len__ = sampler.__len__
```

**Comment by ejguan** [comment]
@bryant1410 
I don't think it would work. Magic functions are attached to class rather than instance.

**Comment by edran** [comment]
@VitalyFedyunin have you made any progress on this? I just got bitten by it, and I suspect it might be a fairly common encounter for anyone that makes custom datasets + uses mypy.

(Also interactions with this issue might not reflect the impact, given it is fairly hard to google it...)

**Comment by VitalyFedyunin** [comment]
> @VitalyFedyunin have you made any progress on this? I just got bitten by it, and I suspect it might be a fairly common encounter for anyone that makes custom datasets + uses mypy.
> 
> (Also interactions with this issue might not reflect the impact, given it is fairly hard to google it...)

Unlikely we will contribute any code to solve this problem for current DataSets as we are putting out resources now to work on torchdata / DataPipes.

**Comment by aaossa** [comment]
I'm facing the same problem. For some reason, even if the custom class inherits from `typing.Sized` the MyPy error doesn't go away. Any ideas on how to patch this on our side in the meantime?

**Comment by ejguan** [comment]
@aaossa Did you add `__len__` method to your custom class?

**Comment by aaossa** [comment]
Yes, I have a custom dataset class that inherits from `Dataset` and implements `__len__` like this:

```python
class CustomDataset(Dataset, Sized):
    # ...
    def __len__(self) -> int:
        return len(self.interactions)
```

In the training loop I'm using a DataLoader, so to access the length of the dataset I use: `len(dataloader.dataset)`. That's the operation that causes the MyPy error:

```raw
main.py:*: error: Argument 1 to "len" has incompatible type "Dataset[Any]"; expected "Sized"  [arg-type]
```

I understand that MyPy complains because the `DataLoader.dataset` attribute has `Dataset[Any]` type, and a `Dataset` is not `Sized`, so that's what I think makes the Sized inheritance useless (`class CustomDataset(Dataset, Sized)`)

**Comment by ejguan** [comment]
Why don't you use `len(dl)`, though?

**Comment by aaossa** [comment]
> Why don't you use `len(dl)`, though?

IIRC that's the number of batches

**Comment by ejguan** [comment]
If you want the size of your dataset, just calling `len(dataset)` is sufficient, right? Or, there is any further detail on why you want to access dataset via attribute in DataLoader

**Comment by aaossa** [comment]
> If you want the size of your dataset, just calling `len(dataset)` is sufficient, right? Or, there is any further detail on why you want to access dataset via attribute in DataLoader

You're right, that works. I'm working on a function that only receives DataLoaders, so that doesn't exactly satisfy what I'm looking for, but it's still progress. I'll have to modify the function or use `# type: ignore` in the `len(dataloader.dataset)` line. Thanks!

**Comment by ejguan** [comment]
@aaossa No worries. Anther workaround might be aliasing it and re-assigning the type to `CustomDataset`. But, you definitely have to add a `type: ignore` when re-assign the type.
```py
ds: CustomDataset = dataloader.dataset  # type: ignore
# Going forward ds should own the type of CustoDataset
length = len(ds)  # No Error should be raised by mypy
```
